rfcs/20200731-Using-Arm-Compute-Library-kernels-on-AArch64/README.md

```diff
@@ -0,0 +1,315 @@
+
+### Background
+
+ArmCL is an open-source library for computer vision and machine learning
+applications and is actively developed by a dedicated team at Arm. The
+development repository is available from
+[mlplatform.org](https://review.mlplatform.org/#/admin/projects/ml/ComputeLibrary),
+and releases are also available on [GitHub](https://github.com/ARM-software/ComputeLibrary).
+Unlike ArmPL, which targets HPC workloads, it has support for bfloat16,
+fixed-point, int8 etc.
+
+At present, ArmCL targets primarily mobile/edge SoCs, but work is ongoing to
+remove bottlenecks to performance on infrastructure scale CPUs and high core
+counts.
+
+ArmCL's library architecture consists of the Core, Runtime and Graph Libraries.
+The Core library contains low-level optimized implementations of the algorithms
+while the Runtime library is a wrapper to the Core that provides memory
+management and multi-threading interfaces. The graph library provides a
+high-level API that can be used to easily chain neural network layers.
+Further details are
+in the Compute Library [documentation](https://github.com/ARM-software/ComputeLibrary/wiki/Documentation).
+
+This initial implementation will demonstrate the integration at the runtime
+level of ArmCL, using an Arm GEMM-based 2D convolution in the forward
+propagation phase. This means that, in the inference regime, convolution calls
+will be supported by ArmCL, and in the training regime only the forward pass
+will be supported; the backward pass will be computed using the existing
+reference implementations.
+
+
+## Proposal
+
+The source code for the initial implementation integrating oneDNN and ArmCL is
+available [here](https://github.com/diaena/oneDNN/tree/armcl-integration-experimental).
+The key changes required are listed below:
+
+- Adding a dependency on ArmCL at the build stage;
+- Adding AArch64 specific code inside a new directory `src/cpu/aarch64`;
+- Adding a `CPU_INSTANCE_AARCH64` macro;
+- Adding ArmCL operations, integrated at runtime-level.
+
+
+### Adding a dependency on ArmCL at the build stage:
+
+To build oneDNN with ArmCL, it must be assumed that a pre-built ArmCL
+library is available, built from the [ArmCL
+sources](https://github.com/ARM-software/ComputeLibrary). Additional
+CMake modules will be needed (`FindARMCL.cmake`, `ArmCL.cmake`) which will only
+impact AArch64 builds. The following steps will be required to get a oneDNN and
+ArmCL build:
+
+- The user, having built ArmCL, will need to set an environment variable
+specifying ArmCL's root directory:
+`export ARMCL_ROOT_DIR=/path/to/ComputeLibrary`;
+
+- Addition of a CMake option to enable ArmCL builds: `DNNL_AARCH64_USE_ARMCL=1`,
+for example, `cmake --verbose .. -DDNNL_AARCH64_USE_ARMCL=1`;
+
+- For a debug build, the user will be required to separately build the debug
+version of ArmCL, which will be entirely independent of oneDNN's CMake flags.
+
+
+### Adding AArch64 specific code inside a new directory `src/cpu/aarch64`:
```

Comment:
Does this plan introduce any code to this repository that is not covered by [THIRD-PARTY-PROGRAMS](https://github.com/oneapi-src/oneDNN/blob/master/THIRD-PARTY-PROGRAMS)?

Reply:
The additional code introduced into `src/cpu/aarch64` would carry the standard Apache 2.0 licence, and an Arm (C) header - `Copyright 2020 Arm Ltd. and affiliates`.

Being a contribution to the oneDNN project, under the established licensing terms, and not a separate project included within oneDNN, it hasn't been listed in the THIRD-PARTY-PROGRAMS. But if that is required in this instance, then we can cover the contents of `src/cpu/aarch64` in THIRD-PARTY-PROGRAMS.

Reply:
Thanks for confirming. No changes in THIRD-PARTY-PROGRAMS are necessary.